### DISCUSSION TITLE
Why does `tokio::sync::oneshot` implement cooperative scheduling?

### DISCUSSION THREAD

**Discussion by olix0r** [tokio-rs/tokio#4776] [Q&A]
There's been a [long-standing bug in Tower](https://github.com/tower-rs/tower/issues/415) that we recently tracked down to some very surprising behavior in `tokio::sync::oneshot`. Basically, we have an assumption that if a `oneshot::Sender` has sent a value, then it must be observable by the `oneshot::Receiver`. But this assumption is violated by [Tokio's cooperative scheduling](https://docs.rs/tokio/latest/tokio/task/index.html#cooperative-scheduling), which can sporadically cause a `oneshot::Receiver::poll` to return Pending even though a value has been sent. The [solution](https://github.com/tower-rs/tower/pull/668) to this issue is to wrap `oneshot::Receiver` with `tokio::task::unconstrained` so that its polls are not impacted by budgeting. This behavior is subtle, implicit, and [undocumented](https://docs.rs/tokio/1.19.2/tokio/sync/oneshot/index.html)--it seems extremely likely that there are other similar bugs lurking anywhere a `oneshot` is used for synchronization.

In my view, this dramatically undermines the utility of `oneshot` as a synchronization primitive.

While I understand the goals of cooperative scheduling task budgeting, I don't understand what this approach protects against in the case of `oneshot`. A receiver:
* can only return ready once
* should not be prone to spurious wake-ups -- i.e. a receiver should not need to return pending more than once in most uses.

What problems would arise if we changed `oneshot`'s behavior to stop participating in task budgeting?

**Comment by olix0r** [comment]
Here's a [minimal example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f977b87e69bd22862d69478c9c3eb3a8) of the issue for those unfamiliar with this:

```rust
#[tokio::main(flavor = "current_thread")]
async fn main() {
    for i in 1..=129 {
        let (tx, rx) = tokio::sync::oneshot::channel();
        tx.send(()).unwrap();
        tokio::select! {
            biased;
            _ = rx => {}
            _ = futures::future::ready(()) => panic!("should not hit this: {} iters", i),
        }
    }
}
```

```text
thread 'main' panicked at 'should not hit this: 129 iters', src/main.rs:12:47
```

**Comment by hawkw** [comment]
> While I understand the goals of cooperative scheduling task budgeting, I don't understand what this approach protects against in the case of `oneshot`. A receiver:
> 
>  * can only return ready once
> 
>  * should not be prone to spurious wake-ups -- i.e. a receiver should not need to return pending more than once in most uses.
> 
> 
> What problems would arise if we changed `oneshot`'s behavior to stop participating in task budgeting?

I believe the reason that polling a `oneshot::Receiver` calls `poll_proceed` is not because a single oneshot is responsible for a large number of polls, but because a task may be polling a very large number of _different_ completed oneshots in a loop or something. 

It's certainly possible that it's not desirable for oneshots to do this, but I'm pretty sure that was the motivation.

**Reply by olix0r** [reply]
@Darksonn and @carllerche explained this a bit more on discord. The real issue is that `oneshot` only really guarantees that the receiver will observe values _at some point in the future_ after a sender sends a value. There are no atomicity/visibility guarantees, even when using `tokio::task::unconstrained`.  To put it simply: `oneshot` is not a synchronization primitive, it deals only in message-passing. A single task that sends a value may not necessarily be able to (immediately) observe that value.

We probably should make these constraints much clearer in the oneshot documentation.

I also suggested that it may be beneficial to have a runtime configuration for tests that enforces much more aggressive (random?) task yielding so that these behaviors are easier to observe in development.

**Reply by barafael** [reply]
> I also suggested that it may be beneficial to have a runtime configuration for tests that enforces much more aggressive (random?) task yielding so that these behaviors are easier to observe in development.

https://crates.io/crates/madsim-tokio

**Comment by BratSinot** [comment]


**Reply by Darksonn** [reply]
Do not use the answer box to post questions. Post a new question instead.
